### PR TITLE
Fixes to make azure iot middleware for freertos to work with ESP-IDF v5.0 and Mbed-TLS v3

### PR DESCRIPTION
**Pull request by ewertons**
- Updated library log tracing arguments to match expected data types
  The gcc-extensa compiler in IDF 5 is much more strict and fails if the arguments to log functions (which use VA ARGS) are incompatible with the format string.

- Update azure_iot_jws_mbedtls.c to work with MBED-TLS v3

  Please see this for the differences between mbed-tls v2 and v3:
  https://github.com/Mbed-TLS/mbedtls/blob/development/docs/3.0-migration-guide.md#remove-the-mode-parameter-from-rsa-functions

  To provide compatibility of the ADU JWS code to both mbed-tls v2 and v3 (and also to better comply with RFC 3447 
  https://www.rfc-editor.org/rfc/rfc3447#section-5.1.2), function mbedtls_rsa_pkcs1_verify is now used instead of 
  mbedtls_rsa_pkcs1_decrypt since the device has only access to the public key.

Tested all ESP32 samples (esp32, azureiotkit, adu) using ESP-IDF versions 4.4.3 and 5.

### CODE REVIEW

File: ports/mbedTLS/azure_iot_jws_mbedtls.c [hunk 1]
```diff
@@ -224,7 +225,7 @@ static AzureIoTResult_t prvJWS_RS256Verify( uint8_t * pucInput,
 
     if( lMbedTLSResult != 0 )
     {
-        AZLogError( ( "[JWS] mbedtls_rsa_import_raw res: %i", lMbedTLSResult ) );
+        AZLogError( ( "[JWS] mbedtls_rsa_import_raw res: %i", ( int ) lMbedTLSResult ) );
```
Comment (vaavva) [inline ports/mbedTLS/azure_iot_jws_mbedtls.c:228]: should all the casts in this PR be int_32's/uint_32's to be more explicit?

**Review by vaavva — APPROVED** [review]

File: ports/mbedTLS/azure_iot_jws_mbedtls.c (hunk 1 above)
Comment (ewertons) [reply ports/mbedTLS/azure_iot_jws_mbedtls.c:228]: Since that's logging, I'm not so particular about standardizing those. But I will if we have a majority of votes :)

**Review by JoseSanchez7 — APPROVED** [review]
I have essentially the same question as Valerie. Does your choosing unsigned ints in certain context have to do with their being able to hold larger positive numbers without extra memory? Or is it just for the sake of diversity and inclusion?

File: source/azure_iot.c [hunk 2]
```diff
@@ -36,7 +36,7 @@ static void prvAzureIoTLogListener( az_log_classification xClassification,
     /* In case logs are stripped out, suppress unused parameter error. */
     ( void ) xMessage;
 
-    AZLogInfo( ( "%.*s", az_span_size( xMessage ), az_span_ptr( xMessage ) ) );
```
Comment (danewalton) [inline source/azure_iot.c:39]: lets use the `int32_t` and `uint32_t` instead of `int` and `unsigned int` just to stick with the code style of this and FreeRTOS (if possible).

File: ports/mbedTLS/azure_iot_jws_mbedtls.c [hunk 3]
```diff
@@ -242,45 +243,39 @@ static AzureIoTResult_t prvJWS_RS256Verify( uint8_t * pucInput,
...
         return xResult;
     }
 
-    /* TODO: remove this once we have a valid PKCS7 parser. */
-    shaMatchResult = memcmp( pucBuffer + azureiotjwsPKCS7_PAYLOAD_OFFSET, pucShaBuffer, azureiotjwsSHA256_SIZE );
+    #if MBEDTLS_VERSION_NUMBER >= 0x03000000
+        lMbedTLSResult = mbedtls_rsa_pkcs1_verify( &ctx, MBEDTLS_MD_SHA256, azureiotjwsSHA256_SIZE, pucBuffer, pucSignature );
+    #else
+        lMbedTLSResult = mbedtls_rsa_pkcs1_verify( &ctx, NULL, NULL, MBEDTLS_RSA_PUBLIC, MBEDTLS_MD_SHA256, azureiotjwsSHA256_SIZE, pucBuffer, pucSignature );
+    #endif
```
Comment (danewalton) [inline ports/mbedTLS/azure_iot_jws_mbedtls.c:265]: I think this works temporarily as the PKCS7 structure which ADU uses is simple. But if they decide to make it more complex (using other signed attributes), I think this could break.

[RFC 3852 Section 5.4](https://www.rfc-editor.org/rfc/rfc3852#section-5.4)

> The message digest calculation process computes a message digest on
   either the content being signed or the content together with the
   signed attributes.

I believe PKCS1 is only the message digest so we're lucky here. 

TLDR: Can we keep the `TODO` so we can hopefully change it to PKCS7 in the future if/when they add one?

**Review by danewalton — APPROVED** [review]

File: ports/mbedTLS/azure_iot_jws_mbedtls.c (hunk 1 above)
Comment (ewertons) [reply ports/mbedTLS/azure_iot_jws_mbedtls.c:228]: Ok, I'll make the changes.

File: source/azure_iot.c (hunk 2 above)
Comment (ewertons) [reply source/azure_iot.c:39]: Ok, I'll update the code.

File: ports/mbedTLS/azure_iot_jws_mbedtls.c (hunk 3 above)
Comment (ewertons) [reply ports/mbedTLS/azure_iot_jws_mbedtls.c:265]: We need to make sure the ADU team gets a feedback about this design, so they take it into consideration. Paraphrasing Cristian, I am surprised the Crypto Board did not raise this flag when they got their design reviewed.

File: source/azure_iot.c (hunk 2 above)
Comment (ewertons) [reply source/azure_iot.c:39]: Done.

File: ports/mbedTLS/azure_iot_jws_mbedtls.c (hunk 1 above)
Comment (ewertons) [reply ports/mbedTLS/azure_iot_jws_mbedtls.c]: Done.

**Comment by ewertons** [comment]
Related issue: https://github.com/Azure/azure-iot-middleware-freertos/issues/301
